[doc] Improve Pulsar security-tls-transport (#5027)
apache/pulsar: site2/docs/security-tls-transport.md

> ## TLS Overview

Reviewer: ```suggestion
## TLS overview
```

By default, Apache Pulsar clients communicate with the Apache Pulsar service in plain text. This means that all data is sent in the clear. You can use TLS to encrypt this traffic to protect the traffic from the snooping of a man-in-the-middle attacker.

You can also configure TLS for both encryption and authentication. Use this guide to configure just TLS transport encryption and refer to [here](security-tls-authentication.md) for TLS authentication configuration. Alternatively, you can use [another authentication mechanism](security-athenz.md) on top of TLS transport encryption.

> Note that enabling TLS may impact the performance due to encryption overhead.

## TLS concepts

> TLS is a form of [public key cryptography](https://en.wikipedia.org/wiki/Public-key_cryptography). Using a key pairs consisting of a public key and a private key can perform the encryption. The public key encrpyts the messages and the private key decrypts the messages.

Reviewer: ```suggestion
TLS is a form of [public key cryptography](https://en.wikipedia.org/wiki/Public-key_cryptography). Using key pairs consisting of a public key and a private key can perform the encryption. The public key encrypts messages and the private key decrypts messages.
```

To use TLS transport encryption, you need two kinds of key pairs, **server key pairs** and a **certificate authority**.

You can use a third kind of key pair, **client key pairs**, for [client authentication](security-tls-authentication.md).

You should store the **certificate authority** private key in a very secure location (a fully encrypted, disconnected, air gapped computer). As for the certificate authority public key, the **trust cert**, you can freely shared it.

For both client and server key pairs, the administrator first generates a private key and a certificate request, then uses the certificate authority private key to sign the certificate request, finally generates a certificate. This certificate is the public key for the server/client key pair.

For TLS transport encryption, the clients can use the **trust cert** to verify that the server has a key pair that the certificate authority signed when the clients are talking to the server. A man-in-the-middle attacker does not have access to the certificate authority, so they couldn't create a server with such a key pair.

> For TLS authentication, the server uses the **trust cert** to verify that the client has a key pair that the certificate authority signed. The Common Name of the **client cert** is then used as the client's role token (see [Overview](security-overview.md)).

Reviewer: ```suggestion
For TLS authentication, the server uses the **trust cert** to verify that the client has a key pair that the certificate authority signed. The common name of the **client cert** is then used as the client's role token (see [Overview](security-overview.md)).
```

> ## Creat TLS Certificates

Reviewer: ```suggestion
## Creat TLS certificates
```

Creating TLS certificates for Pulsar involves creating a [certificate authority](#certificate-authority) (CA), [server certificate](#server-certificate), and [client certificate](#client-certificate).

> Follow the abridged guide below to set up a certificate authority. You can also refer to plenty of resources on the internet for a more detailed guide. We recommend [this guide](https://jamielinux.com/docs/openssl-certificate-authority/index.html) for your detailed reference.

Reviewer: ```suggestion
Follow the guide below to set up a certificate authority. You can also refer to plenty of resources on the internet for a more detailed guide. We recommend [this guide](https://jamielinux.com/docs/openssl-certificate-authority/index.html) for your detailed reference.
```

### Certificate authority

> First, create the certificate for the CA. You can use CA to sign both the broker and client certificates. This ensures that each party will trust the others. You should store CA in a very secure location (ideally completely disconnected from networks, air gapped, and fully encrypted).

Reviewer: ```suggestion
1. Create a certificate for the CA. 

    You can use the CA to sign both the broker and client certificates. This ensures that each party will trust the others. You should store the CA in a very secure location (ideally completely disconnected from networks, air-gapped and fully encrypted).
```

> Next, create a directory for your CA using the follwing command, and place [this openssl configuration file](https://github.com/apache/pulsar/tree/master/site2/website/static/examples/openssl.cnf) in the directory. You may want to modify the default answers for company name and department in the configuration file. Export the location of the CA directory to the environment variable, CA_HOME. The configuration file uses this environment variable to find the rest of the files and directories that the CA needed.

Reviewer: ```suggestion
2. Create a directory for your CA using the following command and place [this OpenSSL configuration file](https://github.com/apache/pulsar/tree/master/site2/website/static/examples/openssl.cnf) in the directory. 

    You may want to modify the default answers for company name and department in the configuration file. Export the location of the CA directory to the environment variable, `CA_HOME`. The configuration file uses this environment variable to find the rest of the files and directories that the CA needed.
```

$ export CA_HOME=$(pwd)
```

> Then, run the commands below to create the necessary directories, keys and certs.

Reviewer: ```suggestion
3. Run the commands below to create necessary directories, keys, and certs.
```

$ chmod 444 certs/ca.cert.pem
```

> After you answer the question prompts, CA-related files are stored in the `./my-ca` directory. Within that directory:

Reviewer: ```suggestion
4. After you answer the question prompts, CA-related files are stored in the `./my-ca` directory. Within that directory:
```

* `certs/ca.cert.pem` is the public certificate. This public certificates is meant to be distributed to all parties involved.
* `private/ca.key.pem` is the private key. You only need it when you are signing a new certificate for either broker or clients and you must safely guard this private key.

### Server certificate

Once you have created a CA certificate, you can create certificate requests and sign them with the CA.

The following commands ask you a few questions and then create the certificates. When you are asked for the common name, you should match the hostname of the broker. You could also use a wildcard to match a group of broker hostnames, for example `*.broker.usw.example.com`. This ensures that multiple machines can reuse the same certificate.

> #### Tips
> 
> Sometimes matching the hostname is not possible or makes no sense,
> such as when you creat the brokers with random hostnames, or you
> plan to connect to the hosts via their IP. In these cases, you 
> should configure the client to disable TLS hostname verification. For more
> details, you can see [the host verification section in client configuration](#hostname-verification).

> First, generate the key using the command below.

Reviewer: ```suggestion
1. Generate the key using the command below.
```

```bash
$ openssl genrsa -out broker.key.pem 2048
```

The broker expects the key to be in [PKCS 8](https://en.wikipedia.org/wiki/PKCS_8) format, so run the following command to convert it.

```bash
$ openssl pkcs8 -topk8 -inform PEM -outform PEM \
      -in broker.key.pem -out broker.key-pk8.pem -nocrypt
```

> Next, use the follwing command to generate the certificate request...

Reviewer: ```suggestion
2. Use the following command to generate the certificate request.
```

```bash
$ openssl req -config openssl.cnf \
      -key broker.key.pem -new -sha256 -out broker.csr.pem
```

> ... and sign it with the certificate authority by running the command below.

Reviewer: ```suggestion
3. Sign the certificate request with the certificate authority by running the command below.
```
